src/lib.rs

```diff
@@ -813,26 +920,54 @@ where
       return Err(NovaError::ProofVerifyError);
     }
 
-    // fold the secondary's running instance with the last instance to get a folded instance
-    let f_U_secondary = self.nifs_secondary.verify(
+    // fold secondary U/W with secondary u/w to get Uf/Wf
+    let r_Uf_secondary = self.nifs_Uf_secondary.verify(
       &vk.ro_consts_secondary,
       &scalar_as_base::<E1>(vk.pp_digest),
       &self.r_U_secondary,
       &self.l_u_secondary,
     )?;
 
+    // fold Uf/Wf with random inst/wit to get U1/W1
+    let r_Un_secondary = self.nifs_Un_secondary.verify(
+      &vk.ro_consts_secondary,
+      &scalar_as_base::<E1>(vk.pp_digest),
+      &r_Uf_secondary,
+      &self.l_ur_secondary,
+    )?;
+
+    // fold primary U/W with random inst/wit to get U2/W2
+    let r_Un_primary = self.nifs_Un_primary.verify(
+      &vk.ro_consts_primary,
+      &vk.pp_digest,
+      &self.r_U_primary,
+      &self.l_ur_primary,
+    )?;
+
+    // derandomize/unblind commitments
+    let derandom_r_Un_primary = r_Un_primary.derandomize(
+      &vk.vk_primary_commitment_key,
+      &self.primary_blind_r_W,
```

Comment:
Nit: `blind_r_W_primary` rather than `primary_blind_r_W` to be consistent with the rest of the names?

Similar comment for other instances.